README
NAME
    Bot::CPAN - provides CPAN services via IRC

SYNOPSIS
      use Bot::CPAN;

      my $bot = Bot::CPAN->new(
          channels => ['#cpan'],
          servers => ['irc.perl.org', 'irc.freenode.net'],
          port => 6667,
          nick => 'cpantest',
          alt_nicks => ['cpantest2', 'cpantest3'],
          username => 'cpantest',
          name => 'cpantest',
          nickserv_password => 'top_secret_password',
          ignore_list => [qw(purl)],
          news_server => 'nntp.perl.org',
          group => 'perl.cpan.testers',
          reload_indices_interval => 300,
          inform_channel_of_new_uploads => 60,
          inform_channel_of_new_ratings => 60,
          debug => 0,
          search_max_results => 30,
          adminhost =>
              qr/Fox!afoxson\@pool-141-158-116-119\.pitt\.east\.verizon\.net/,
          policy => { # See 'POLICY CONTROL MECHANISM', in the POD, below
              '#cpan' => {             # allow channel 'cpan upload' informs,
                  allow => qr/^POE-/i, # for #cpan, only if it matches ^POE-,
                  deny => qr/.+/,      # and deny everything else
              },
          },
      );
      $bot->run();

      # OR, see the 'cpanbot' script for an alternative

DESCRIPTION
    Bot::CPAN is a POE based distribution that allows individuals on IRC to
    query the CPAN and other perl resources in a great number of different
    ways. Bot::CPAN will also automatically inform the channels of new
    uploads to the CPAN, and new ratings of CPAN distributions.

IRC EXAMPLES
      cpan: recent
      cpan: top10
      cpan: id_name for FOX
      cpan: mod_desc for Test::Reporter
      cpan: dist_version for Bot-CPAN

METHODS
    * new
        This constructor returns a Bot::CPAN object, and is inherited from
        Bot::CPAN::BasicBot. It will accept named parameters for: channels,
        servers, port, nick, alt_nicks, username, name, ignore_list, store,
        and log. Bot::CPAN extends the Bot::CPAN::BasicBot constructor to
        accept the following named parameters: news_server, group,
        reload_indices_interval, inform_channel_of_new_uploads,
        inform_channel_of_new_ratings, debug, search_max_results,
        nickserv_password, policy and adminhost.

    * run
        Fires up the bot.

CONSTRUCTOR OPTIONS
    * adminhost
        Specifies a regular expression that will be matched against
        userhosts for commands that require administrative access.

    * alt_nicks
        Alternate nicks that this bot will be known by. These are not nicks
        that the bot will try if its main nick is taken, but rather other
        nicks that the bot will recognize if it is addressed in a public
        channel as the nick. This is useful for bots that are replacements
        for other bots...e.g, your bot can answer to the name "infobot: "
        even though it isn't really.

    * channels
        The channels we're going to connect to.

    * ignore_list
        The list of irc nicks to ignore public messages from (normally other
        bots.) Useful for stopping bot cascades.

    * debug
        Enable or disable bugging. 1 or 0.

    * group
        NNTP group to retrieve articles from. This group should be where the
        cpan upload emails are sent.

    * inform_channel_of_new_uploads
        Number of seconds between checks for new CPAN uploads.

    * inform_channel_of_new_ratings
        Number of seconds between checks for new CPAN ratings.

    * name
        The name that the bot will identify itself as.

    * nickserv_password
        Password for IRC networks that use NICKSERV.

    * news_server
        The NNTP server to retrieve articles from.

    * nick
        The nick we're going to use.

    * policy
        Defines the policy control mechanism.

    * port
        The port we're going to use.

    * reload_indices_interval
        Number of seconds between reloading of CPAN indices.

    * search_max_results
        Maximum numer of results to return via the 'search' command.
        (semi-deprecated)

    * servers
        The servers that the bot should attempt to connect to. One will be
        chosen at random for every connect, or reconnect.

    * username
        The username we'll claim to have at our ip/domain.

COMMANDS
    * botsnack
        gives the bot a snack

    * config
        Shows the bot's configuration (admin only)

    * help
        provides instruction on how to use this bot

    * proxy
        proxies an emote (admin only)

    * recent
        shows last ten distributions uploaded to the CPAN

    * status
        retrieves the status of the bot

    * top10
        top 10 contributors to CPAN

    * id_dists
        shows all distributions of an author

    * id_email
        retrieves an email from a PAUSE ID

    * id_name
        retrieves a name from a PAUSE ID

    * mod_chapter
        retrieves the chapter of a module

    * mod_desc
        retrieves the description of a module

    * mod_dist
        retrieves the distribution of a module

    * mod_docurl
        retrieves the documentation URL of a module

    * mod_dslip
        retrieves the DSLIP of a module

    * mod_id
        retrieves the PAUSE ID of a module

    * mod_language
        retrieves the language of a module

    * mod_license
        retrieves the license of a module

    * mod_rturl
        retrieves the RT URL of a module

    * mod_stage
        retrieves the stage of a module

    * mod_search
        search for a module

    * mod_support
        retrieves the support level of a module

    * mod_style
        retrieves the style of a module

    * mod_version
        retrieves the version of a module

    * dist_chapter
        lists the chapter of a distribution

    * dist_date
        retrieves the release date of a distribution

    * dist_desc
        retrieves description of a distribution

    * dist_dlurl
        retrieves the download URL of a distribution

    * dist_filename
        retrieves the filename of a distribution

    * dist_id
        retrieves PAUSE ID of a distribution

    * dist_mods
        lists all modules in a distribution

    * dist_path
        retrieves the path to a distribution

    * dist_ratings
        retrives ratings of a distribution

    * dist_search
        earch for a distribution

    * dist_size
        retrieves the size of a distribution

    * dist_tests
        shows tests results for a distribution

    * dist_url
        retrieves URL of a distribution

    * dist_version
        retrieves the version of a distribution

ATTRIBUTES
    * :Admin
        Indicates that this command is an admin command. A user will not be
        able to execute commands marked with this attribute unless their
        usermask matches the adminhost regex specified in the constructor.

    * :Args
        Indicates the nature of the arguments sent to this command. This
        attribute takes an argument of either 'required', 'optional', or
        'refuse'. The user will then get an error message if they attempt to
        use a command in a manner inconsistent with its intended use.

    * :Fork
        Indicates that this command should be forked off. This should be
        used only for commands that take a long time to execute (like
        'tests'). Essentially, forking off a long running command will
        prevent the execution from blocking the bot.

    * :Help
        Defines the help message for this command, which will be available
        via: /msg bot help <command>.

    * :LowPrio
        Indicates that this command's data should be returned to the user
        with a low priority. This should be used only for commands that
        return a lot of discrete chunks of data back to the user (like
        'tests'). This will prevent the returning of a lot of data from
        blocking the bot.

    * :Private
        Indicates that this command can be executed via a /msg. This
        attribute takes an argument of either 'notice', or 'privmsg',
        indicating that manner in which the data should be returned to the
        user.

    * :Public
        Indicates that this command can be executed publically from within a
        channel. This attribute takes an argument of either 'notice', or
        'privmsg', indicating that manner in which the data should be
        returned to the user.

HOW TO WRITE A NEW COMMAND
    Simple really. Define a subroutine, in CPAN.pm, and add whatever
    attributes you'd like to it's signature (see above). You will be passed
    three arguments, the referrent object, the event hashref, and the
    command's actual argument. To get data back to the user, simply use
    $self->print(). Its first argument should be the event hashref and the
    other argument is whatever you want to return to the user.

POLICY CONTROL MECHANISM
    The policy control mechanism consults the policy hashref specified in
    the constructor. The policy hashref should contain channel hashrefs. The
    channel hashref may contain both 'allow', and 'deny' key/value pairs.
    The values of the keys are regex's which specify which channel 'cpan
    upload' informs are to be allowed or denied. The search stops at the
    first match.

    This is a simplistic and minimal policy control mechanism that is
    directly based on HOSTS_ACCESS(5) (/etc/hosts.(allow|deny)).

    A channel 'cpan upload' inform will be allowed when the text of an
    inform matches the allow regex.

    Otherwise, a channel 'cpan upload' inform will be denied when the text
    of an inform matches the deny regex.

    Otherwise, a channel 'cpan upload' inform will be allowed. If a channel
    doesn't have a specified policy, all channel 'cpan upload' informs will
    be allowed.

    You will be matching against strings that look like the below. Bear this
    firmly in mind when you create regex's to match them (see 'RATINGS'):

      Test-Reporter-1.19 (+++++) by AFOXSON
      POE-0.25 (+++++) by RCAPUTO
      Games-Cryptoquote-1.30 (++++ ) by BOBO

RATINGS
    An upload inform will look something like:

      Test-Reporter-1.19 (+++++) by AFOXSON

    What is up with all of the +'s? The plus signs within the parens
    represent the distributions average rounded rating from
    cpanratings.perl.org (aka karma). Possible ratings are 1 through 5. The
    area between the parens is fixed-width which means if a distribution has
    an average rounded ratinging of '2' if will be padded with 3 spaces.

    It used to be that (unknown) would be displayed if the bot cannot
    connect to the data source, (unrated) would be displayed if the module
    has yet to receieve any ratings, and (error) would be displayed if
    something went wrong. However, after some discussion on #perl@MagNET,
    the karma representation will be omitted unless there is an actual karma
    rating to display.

    There were competing "karma represetations" floated on #perl@MagNET,
    such as "letter-grades", and "descriptive words" but in the end, the
    incremement style got the popular vote.

CAVEATS
    If you expect the bot to have very recent CPAN data, be sure that
    reindexing is set to get indexes directly from ftp.funet.fi. Even then,
    those indices are only updated on the funet end about once an hour or
    so.

    Bot::CPAN::BasicBot's alt_nicks doesn't do what you think. It's NOT for
    specifying alternate nicks to use for connecting to IRC if the one you
    chose is taken. It's used for specifying nicks that you'd like the bot
    to also respond for, as if the real nick was addressed. This is a relic
    from the bundling of Bot::BasicBot.

BUGS
    If you happen to find one please email me at afoxson@pobox.com. Thank
    you. Or, better yet, report it on RT.

TODO
      - Add support for "throttling" (suggested by Spoon) i.e., if several uploads
        occur at the same time, pace the reporting of such to a pre-determined
        number of upload notification per minute
      - Consider re-adding 'modulelist'
      - Consider re-adding 'details'
      - Consider re-supporting search_max_results
      - Improve 'recent' so it has a full recent list on connect
      - Improve memory profile
      - Investigate forking reindexing and checking for new uploads/ratings

COPYRIGHT
      Copyright (c) 2003 Adam J. Foxson. All rights reserved.
      Copyright (c) 2003 Casey West. All rights reserved.

LICENSE
    This program is free software; you can redistribute it and/or modify it
    under the terms of the GNU General Public License as published by the
    Free Software Foundation; either version 2 of the License, or (at your
    option) any later version.

    This program is distributed in the hope that it will be useful, but
    WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
    Public License for more details.

SEE ALSO
    * perl
    * HTTP::Request
    * XML::Parser
    * URI
    * LWP::UserAgent
    * Mail::Internet
    * Math::Round
    * Net::NNTP
    * POE
    * Statistics::Descriptive
    * XML::RSS::Parser
    * POE::Component::IRC
    * Text::Wrap
    * Error
    * Class::Phrasebook
    * Attribute::Handlers
    * Storable
    * CPAN::DistnameInfo
    * Compress::Zlib
    * File::Listing
    * Sort::Versions
    * Encode
    * Config::Auto
    * Getopt::Long

AUTHORS
    Adam J. Foxson <afoxson@pobox.com>, with patches from Casey West
    <cwest@cpan.org> to support the latest POE versions, Randal Schwartz
    <merlyn@stonehenge.com> to support NNTP retrieval of CPAN uploads (as
    opposed to the old way of doing it via mailbox polling), and Rocco
    Caputo <troc+cpan@pobox.com> that solved early-on blocking issues, and
    got the prioritized events patch into the P::C::I core. Special thanks
    goes out to Iain Truskett for diligent testing and the suggestion of
    many spiffy features.



Changes
Revision history for Perl extension Bot::CPAN.

1.11  Tue Jul 04 20:00:00 2006
      - Oops, forgot to regenerate README!
      - See changes in 1.10 for the latest updates

1.10  Tue Jul 04 19:00:00 2006
      - Added exception handling classes for certain events
      - Removed dependency on CPANPLUS
      - Added dependency on CPAN::Search::Lite::Info for indexing
      - Merged in latest changes from Bot::BasicBot
      - Accounted for recent changes in POE and POE::Component::IRC
      - Resolved an issue with Attribute::Handlers (Invalid CODE attributes)
      - Resolved an issue with Bot::BasicBot (Can't encode object)
      - Resolved an issue where forked off operations wouldn't be reaped
      - Resolved several miscellaneous minor issues
      - The bot will now report on new CPAN ratings in addition to uploads
      - Added new option 'inform_channel_of_new_ratings'
      - Added support for a NICKSERV password, i.e., on freenode
      - Added new option 'nickserv_password'
      - SIGNATURE now based on new key; old key expired
      - Corrected an issue where spurious "1"'s appeared after certain cmds
      - Corrected an issue where users where not addressed in forked cmds
      - Internal cleanup in several areas
      - The bot no longer considers "$botname\s+$command" as being addressed
      - Deprecated the 'reviews' command; cpanratings no longer supports rss here
      - Deprecated the 'wikiurl' command; this resource no longer exists
      - Deprecated the 'readme' command; excessive resource consumption
      - Deprecated the 'package' command; replaced by an alternative
      - Deprecated the 'modules' command; non-sensical command
      - Deprecated the 'modulelist' command; may re-add in the future
      - Deprecated the 'details' command; may re-add in the future
      - Added many new commands; see the documentation
      - Commands now follow a consistent naming convention
      - Added new script that regenerates the indices as a separate process
      - Added more tests
      - Fixed a few ostensibly nasty memory leaks
      - Bot can now (easily) run unprivileged
      - search_max_results semi-deprecated; applicable commands /msg only

1.7  Thu Aug 28 05:28:04 2003
      - Added new test suite 'phrases.t', with 68 tests
      - All messages abstracted out to lib/Bot/CPAN/phrases.xml, thanks to
        Class::Phrasebook
      - 'ratings' & 'reviews' now do auto-module-to-dist translation (per spoon)
      - Added 'mode' stat to 'ratings' (per cwest)
      - Optional 'cpanbot' script is now stable, but experimental. AppConfig
        was dumped in favor of Getopt::Long and Config::Auto (per cwest)
      - Fixed a display bug that added an extraneous space whilst logging events
      - Added optional support for better guessing of distribution descriptions
        via Module::CPANTS
      - Syncronized B::C::BasicBot 0.05 against Bot-BasicBot-0.2; which means
        to say that i've reviewed the diffs and found nothing compelling to
        change :)
      - Fixed a bug with the 'distributions' command
      - Added support for CPANPLUS version 0.043
      - Karma ratings (+++++) for upload notifications will only be shown if
        there is a karma rating *to* be shown, i.e., (unrated) is no longer
        displayed (per schwern)

0.83  Fri Aug 22 19:54:29 2003
      - Changed display format of 'ratings'

0.82  Wed Aug 20 01:58:31 2003
      - Started working on the 'cpanbot' wrapper; it is currently experimental
        and non-functional. Technically, the wrapper is complete, but there is
        a serious problem with using Bot::CPAN with AppConfig that I haven't
        as yet solved *sigh*
      - Applied a patch from dngor and Spoon that fixes a bug with reconnection
      - We now use Statistics::Descriptive and Math::Round to calculate
        cpanratings stats
      - Added 'min', 'max', and 'standard deviation' stats, to the existant
        stats of 'mean', and 'median' for cpanratings stats
      - Added karma-ish representation of the average, rounded rating for any
        given distribution to the upload informs. i.e.:
        CPAN upload: CGI-Wiki-0.46 (++++ ) by KAKE
        Note: There were competing "karma represetations" floated on
              #perl@MagNET, such as "letter-grades", and "descriptive words"
              but in the end, the incremement style got the popular vote

0.79  Fri Aug 15 10:10:06 2003
      - Added command 'reviews'; retrieves reviews from cpanratings.perl.org
      - Added command 'ratings'; retrieves ratings from cpanratings.perl.org

0.76  Sun Aug 10 9:42:21 2003
      - The 'url' command now returns a /dist/ url, instead of an /author/
        one. This is mainly for the benefit of people that run Bot::CPAN off
        of 'delayed mirrors', as it will generate a url that refers to the
        very latest version of any given distribution. If we can't manage to
        work out the proper /dist/ url, it will return an /author/ one
      - Added command 'wikiurl' which returns the url to the module's wiki
        at http://cpan.japh.org/
      - Added command 'docurl' which returns the url to the module's
        documentation on search.cpan.org
      - All of these changes/additions were inspired by Spoon

0.74  Sat Mar 22 23:31:47 2003
      - It's no longer necessary to patch P::C::I for B::C to function. We
        now have a dependency on POE::Component::IRC 2.9 (which now supports
        prioritized messaging; fimmtiu++ dngor++)
      - Integrated changes from Bot::BasicBot 0.05
      - Added 'Last indice reload' stat to 'config'
      - Fixed some issues with addressing
      - Fixed a major issue with bots not correctly reindexing; in order to
        fix the reindexing issue, it was necessary to make reload_indices
        *not* fork
      - Added bot-command 'modulelist', as in:
        'cpan: modulelist contains Test::Reporter?'
      - Changed verbage of 'public only' to 'channel only', and

Makefile.PL
# $Revision: 1.3 $
# $Id: Makefile.PL,v 1.3 2006/07/04 08:04:29 afoxson Exp $

use inc::Module::Install;

name('Bot-CPAN');
version_from('lib/Bot/CPAN.pm');
abstract('provides CPAN services via IRC');
author('Adam J. Foxson <afoxson@pobox.com>');
license('gpl');

requires('HTTP::Request' => '1.40');
requires('XML::Parser' => '2.34');
requires('URI' => '1.35');
requires('LWP::UserAgent' => '2.033');
requires('Mail::Internet' => '1.74');
requires('Math::Round' => '0.05');
requires('Net::NNTP' => '2.23');
requires('POE' => '0.3502');
requires('Statistics::Descriptive' => '2.6');
requires('XML::RSS::Parser' => '4');
requires('POE::Component::IRC' => '4.93');
requires('Text::Wrap' => '2005.082401');
requires('Error' => '0.16');
requires('Class::Phrasebook' => '0.88');
requires('Attribute::Handlers' => '0.78');
requires('Storable' => '2.15');
requires('CPAN::DistnameInfo' => '0.06');
requires('Compress::Zlib' => '1.41');
requires('File::Listing' => '1.15');
requires('Sort::Versions' => '1.5');

features(
	'utf8 decoding of cpanratings data' => [
		-default => 1,
		recommends('Encode' => '2.18'),
	],
	'support for the \'cpanbot\' wrapper script' => [
		-default => 0,
		recommends('Config::Auto' => '0.16'),
		recommends('Getopt::Long' => '2.35'),
	],
);

include('ExtUtils::AutoInstall');
check_nmake();
auto_install();

&Meta->write;
&Build->write if lc($0) eq 'build.pl';
&Makefile->write if lc($0) eq 'makefile.pl';
